Init network

Vincinator opened the pull request:
Hi, 


This PR uses the crate [little-dude/netlink](https://github.com/little-dude/netlink) to initialize network devices

* Added network initialization for the libvirt-pid1 initialization flow
    * Initialize `lo` device
        * assign `::1/128`
        * assign `127.0.0.1/8`
    * Initialize `eth0` device
        * assign `fe80::2/64`
    * Print network info after initialization
* Functions used in above initialization flow are a starting point for other scenarios
* Possibility to create sriov devices for a given network interface

## How to test

As usual, copy paste steps from [hack/README.md](https://github.com/aurae-runtime/auraed/blob/main/hack/README.md):

```
make build-container
make kernel
make initramfs

# create `vm-br0` bridge on your machine:
make network

# run auraed in a VM as pid 1:
make virsh-start virsh-console virsh-stop

# exit VM console with Ctrl+]
```

On a second terminal you are now able to ping the auraed node via
```
# name of network interface "vm-br0" is set up by make network
ping fe80::2%vm-br0 
```

## Screenshot
![2022-09-29_19-54-57_screenshot](https://user-images.githubusercontent.com/2008076/193106981-79010f06-a9a8-41e8-b1b4-a2d9976a7ab4.png)

Vincinator committed:
add basic network init for libvirt+pid1

Vincinator committed:
Merge branch 'main' into init-network

Vincinator committed:
use generic setup for loopback device

Vincinator committed:
remove system folder

krisnova commented on src/init/network.rs:
```diff
@@ -0,0 +1,281 @@
+    .expect("Unable to write file");
+}
+
+pub async fn init_iface(
```
function signatures should denote the IP stack (IPv4) or (IPv6)

```rust
pub async fn init_iface_ipv6();
```

Vincinator replied:
I removed the init_iface functions completely now.
I think the initialization code of a given scenario can directly call the network functions.

krisnova commented on src/init/network.rs:
```diff
@@ -0,0 +1,281 @@
+    let localhost_ipv4: IpNetwork = "127.0.0.1/8".parse().unwrap();
+    let localhost_ipv6: IpNetwork = "::1/128".parse().unwrap();
+
+    if let Err(e) = add_address("lo", localhost_ipv4, handle.clone()).await {
```
move "lo" to constant at top of file as we will likely have multiple loopback names in the future

krisnova commented on src/init/network.rs:
```diff
@@ -0,0 +1,281 @@
+    }
+
+    if let Err(e) = add_address("lo", localhost_ipv6, handle.clone()).await {
+        error!("{:?}", e);
```
include IP stack in error message please 

```rust
        error!("IPv6 Error: {:?}", e); // or similar
```

Vincinator replied:
good idea. Added ipstack in error message in result checking of `add_address_ipv6` and `add_address_ipv4` calls.

The `init_iface` function is removed

krisnova commented on src/init/network.rs:
```diff
@@ -0,0 +1,281 @@
+                                );
+                            }
+                        }
+                        None => {}
```
Should we log here?

Vincinator replied:
yes!

krisnova commented on src/lib.rs:
```diff
@@ -171,30 +176,43 @@ pub struct SystemRuntime {
+        let (connection, handle, _) = new_connection().unwrap();
+        tokio::spawn(connection);
+
+        trace!("configure lo");
```
use "looback" for logs as "lo" is subject to change in future PRs

Vincinator replied:
done

krisnova commented on src/lib.rs:
```diff
@@ -171,30 +176,43 @@ pub struct SystemRuntime {
+        let (connection, handle, _) = new_connection().unwrap();
+        tokio::spawn(connection);
+
+        trace!("configure lo");
```
use "looback" for logs as "lo" is subject to change in future PRs

krisnova reviewed (changes requested):
Please see my notes 👀

Vincinator committed:
Merge branch 'aurae-runtime:main' into init-network

Vincinator committed:
fix: build-container must add clippy

Vincinator committed:
refactor: directly call network functions in init

Vincinator commented:
Implemented your feedback

I removed the `init_iface` functions completely now.
I think the initialization code of a given scenario can directly call the network functions.

krisnova reviewed (approved):
LGTM -- let's keep moving forward
